src/report/filter.rs

```diff
@@ -0,0 +1,137 @@
+use std::path::PathBuf;
+
+use fnv::FnvHashMap;
+
+use crate::fileset::FileKind;
+use crate::report::{ErrorKey, LogLevel, LogReport};
+use crate::token::Loc;
+
+/// Determines whether a given Report should be printed.
+/// If a report is matched by both the blacklist and the whitelist, it will not be printed.
+#[derive(Default)]
+pub struct ReportFilter {
+    /// Whether to log errors in vanilla CK3 files
+    pub show_vanilla: bool,
+    /// Whether to log errors in other loaded mods
+    pub show_loaded_mods: bool,
+    /// Minimum error level to log
+    pub min_level: LogLevel,
+    /// If a whitelist exists, reports must match ALL FilterMatches to be eligible to be printed.
+    pub whitelist: Option<FilterMatches>,
+    /// If a blacklist exists, reports must match NONE of the FilterMatches to be eligible to be printed.
+    pub blacklist: Option<FilterMatches>,
+}
+
+impl ReportFilter {
+    /// Returns true iff the report should be printed.
+    /// A print will be rejected if the report matches at least one of the following conditions:
+    /// - Its Severity or Confidence level is too low.
+    /// - It's from vanilla or a loaded mod and the program is configured to ignore those locations.
+    /// - The filter has a whitelist, and the report doesn't match it.
+    /// - The filter has a blacklist, and the report does match it.
+    pub fn should_print_report(&self, report: &LogReport) -> bool {
+        self.should_print(report.lvl, &report.primary().location, report.key)
+    }
+    pub fn should_print(&self, lvl: LogLevel, location: &Loc, key: ErrorKey) -> bool {
+        if lvl.severity < self.min_level.severity
+            || lvl.confidence < self.min_level.confidence
+        {
+            return false;
+        }
+        let linked_location = Self::get_end_of_link_chain(location);
+        if (linked_location.kind <= FileKind::Vanilla && !self.show_vanilla)
+            || (matches!(linked_location.kind, FileKind::LoadedMod(_)) && !self.show_loaded_mods)
+        {
+            return false;
+        }
+        if let Some(whitelist) = &self.whitelist {
+            if !whitelist.matches(linked_location, key) {
+                return false;
+            }
+        }
+        if let Some(blacklist) = &self.blacklist {
+            if blacklist.matches(linked_location, key) {
+                return false;
+            }
+        }
+        true
+    }
+    /// Check all elements of the loc link chain.
+    /// This is necessary because of cases like a mod passing `CHARACTER = this` to a vanilla
+    /// script effect that does not expect that. The error would be located in the vanilla script
+    /// but would be caused by the mod.
+    fn get_end_of_link_chain(loc: &Loc) -> &Loc {
```

Comment:
Hmm here you just check the top level of the link chain. I don't think that works right.

A common example of blacklist use is if a mod copies vanilla files into the mod, and doesn't want to fix bugs in the copied files. Then the modder adds those vanilla files to the blacklist, and errors located in them should not be reported. Even if the mod called into that file. The modder would not want to add the mod files to the blacklist to cover up these reports, because errors in the modded files need to be caught.

For a similar reason, the no-vanilla check should also check all links in the chain, but with the opposite logic: it should print the report if any of them are modded files. This covers the case where a mod replaces a scripted effect that then causes errors when called from vanilla locations. (Blacklisting should be stronger than this check.)

Reply:
I might be wrong, but I'm pretty sure that's what it always used to do.

Reply:
Never mind, I see now.

Reply:
I realize now that my partially short-circuited recursion was hard to read :) could definitely use a comment